contracts/core/EntryPoint.sol

```diff
@@ -595,42 +599,79 @@ contract EntryPoint is IEntryPoint, StakeManager, NonceManager, ReentrancyGuardT
      * @return validationData                    - The Paymaster's validationData.
      */
     function _validatePaymasterPrepayment(
         uint256 opIndex,
         PackedUserOperation calldata op,
-        UserOpInfo memory opInfo,
-        uint256 requiredPreFund
+        UserOpInfo memory opInfo
     ) internal virtual returns (bytes memory context, uint256 validationData) {
         unchecked {
             uint256 preGas = gasleft();
             MemoryUserOp memory mUserOp = opInfo.mUserOp;
             address paymaster = mUserOp.paymaster;
+            uint256 requiredPreFund = opInfo.prefund;
             if (!_tryDecrementDeposit(paymaster, requiredPreFund)) {
                 revert FailedOp(opIndex, "AA31 paymaster deposit too low");
             }
             uint256 pmVerificationGasLimit = mUserOp.paymasterVerificationGasLimit;
-            try
-            IPaymaster(paymaster).validatePaymasterUserOp{gas: pmVerificationGasLimit}(
-                op,
-                opInfo.userOpHash,
-                requiredPreFund
-            )
-            returns (bytes memory _context, uint256 _validationData) {
-                context = _context;
-                validationData = _validationData;
-            } catch {
-                revert FailedOpWithRevert(opIndex, "AA33 reverted", Exec.getReturnData(REVERT_REASON_MAX_LEN));
-            }
+            (context, validationData) = _callValidatePaymasterUserOp(opIndex, op, opInfo);
             if (preGas - gasleft() > pmVerificationGasLimit) {
                 revert FailedOp(opIndex, "AA36 over paymasterVerificationGasLimit");
             }
         }
     }
 
+    function _callValidatePaymasterUserOp(
+        uint256 opIndex,
+        PackedUserOperation calldata op,
+        UserOpInfo memory opInfo
+    ) internal returns (bytes memory context, uint256 validationData)  {
+        uint256 freePtr = _getFreePtr();
+        bytes memory validatePaymasterCall = abi.encodeCall(
+            IPaymaster.validatePaymasterUserOp,
+            (op, opInfo.userOpHash, opInfo.prefund)
+        );
+        address paymaster = opInfo.mUserOp.paymaster;
+        uint256 paymasterVerificationGasLimit = opInfo.mUserOp.paymasterVerificationGasLimit;
+        bool success;
+        uint256 contextLength;
+        uint256 contextOffset;
+        uint256 maxContextLength;
+        assembly ("memory-safe") {
+            //call and return 3 first words: offset, validation, context-length
+            success := call(paymasterVerificationGasLimit, paymaster, 0, add(validatePaymasterCall, 0x20), mload(validatePaymasterCall), freePtr, 96)
+            validationData := mload(add(freePtr, 32))
+            contextLength := mload(add(freePtr, 64))
+            contextOffset := mload(freePtr)
+            maxContextLength := sub(returndatasize(), 96)
+        }
+        if (!success || contextOffset != 64 || contextLength > maxContextLength) {
```

Comment:
Regarding `contextOffset != 64`, this is not a strict requirement of the ABI encoding as far as I can tell. 

If someone were to implement the paymaster interface this way:
```
function validatePaymasterUserOp(
...
) external returns (bytes memory context, uint256 validationData, uint256 myBirthday);
```

This function would've been valid with Solidity but not with this assmebly, I assume.
Is this so? Should we care about it?

Reply:
At least document the magic numbers, and explain that you're assuming a specific abi encoding of the return value